src/Validation/URLValidationProvider.php

```diff
@@ -0,0 +1,216 @@
+use AMP_Validation_Manager;
+use WP_Error;
+
+/**
+ * URLValidationProvider class.
+ *
+ * @since 2.1
+ */
+final class URLValidationProvider {
+
+	/**
+	 * Key for the transient signaling validation is locked.
+	 *
+	 * @var string
+	 */
+	const LOCK_TRANSIENT = 'amp_validation_locked';
+
+	/**
+	 * Flag to pass to get_url_validation to force revalidation.
+	 *
+	 * @param string
+	 */
+	const FLAG_FORCE_REVALIDATE = 'amp_force_revalidate';
+
+	/**
+	 * Flag to pass to get_url_validation to skip revalidation.
+	 *
+	 * @param string
+	 */
+	const FLAG_NO_REVALIDATE = 'amp_no_revalidate';
+
+	/**
+	 * The total number of validation errors, regardless of whether they were accepted.
+	 *
+	 * @var int
+	 */
+	public $total_errors = 0;
+
+	/**
+	 * The total number of unaccepted validation errors.
+	 *
+	 * If an error has been accepted in the /wp-admin validation UI,
+	 * it won't count toward this.
+	 *
+	 * @var int
+	 */
+	public $unaccepted_errors = 0;
+
+	/**
+	 * The number of URLs crawled, regardless of whether they have validation errors.
+	 *
+	 * @var int
+	 */
+	public $number_crawled = 0;
+
+	/**
+	 * The validation counts by type, like template or post type.
+	 *
+	 * @var array[] {
+	 *     Validity by type.
+	 *
+	 *     @type array $type {
+	 *         @type int $valid The number of valid URLs for this type.
+	 *         @type int $total The total number of URLs for this type, valid or invalid.
+	 *     }
+	 * }
+	 */
+	public $validity_by_type = [];
+
+	/**
+	 * Locks validation.
+	 */
+	private function lock() {
+		set_transient( self::LOCK_TRANSIENT, 'locked', $this->get_lock_timeout() );
+	}
+
+	/**
+	 * Unlocks validation.
+	 */
+	private function unlock() {
+		delete_transient( self::LOCK_TRANSIENT );
+	}
+
+	/**
+	 * Returns whether validation is currently locked.
+	 *
+	 * @return boolean
+	 */
+	public function is_locked() {
+		return get_transient( self::LOCK_TRANSIENT );
+	}
+
+	/**
+	 * Provides the length of time, in seconds, to lock validation when this runs.
+	 *
+	 * @return int
+	 */
+	private function get_lock_timeout() {
+		/**
+		 * Filters the length of time to lock URL validation when a process starts.
+		 *
+		 * @param int $timeout Time in seconds. Default 300 seconds.
+		 */
+		return apply_filters( 'amp_validation_lock_timeout', 5 * MINUTE_IN_SECONDS );
+	}
+
+	/**
+	 * Runs a callback with a lock set for the duration of the callback.
+	 *
+	 * @param callable $callback Callback to run with the lock set.
+	 * @return mixed  WP_Error if a lock is in place. Otherwise, the result of the callback or void if it doesn't return anything.
+	 */
+	public function with_lock( $callback ) {
+		if ( $this->is_locked() ) {
+			return new WP_Error(
+				'amp_url_validation_locked',
+				__( 'URL validation cannot start right now because another process is already validating URLs. Try again in a few minutes.', 'amp' )
+			);
+		}
+
+		$this->lock();
+		$result = call_user_func( $callback );
```

Comment:
This should just execute the callable (which is also much faster):
```suggestion
		$result = $callback();
```